backends/ref/ceed-ref-basis.c

```diff
@@ -306,7 +306,112 @@ static int CeedBasisApply_Ref(CeedBasis basis, CeedInt num_elem,
 }
 
 //------------------------------------------------------------------------------
-// Basis Create Non-Tensor
+// Basis Apply Hdiv
+//------------------------------------------------------------------------------
+static int CeedBasisApplyHdiv_Ref(CeedBasis basis, CeedInt num_elem,
+                                  CeedTransposeMode t_mode, CeedEvalMode eval_mode,
+                                  CeedVector U, CeedVector V) {
+  int ierr;
+  Ceed ceed;
+  ierr = CeedBasisGetCeed(basis, &ceed); CeedChkBackend(ierr);
+  CeedInt dim, num_comp, num_nodes, num_qpts;
+
+  ierr = CeedBasisGetDimension(basis, &dim); CeedChkBackend(ierr);
+  ierr = CeedBasisGetNumNodes(basis, &num_nodes); CeedChkBackend(ierr);
+  ierr = CeedBasisGetNumQuadraturePoints(basis, &num_qpts); CeedChkBackend(ierr);
+  ierr = CeedBasisGetNumComponents(basis, &num_comp); CeedChkBackend(ierr);
+  CeedTensorContract contract;
+  ierr = CeedBasisGetTensorContract(basis, &contract); CeedChkBackend(ierr);
+  const CeedInt add = (t_mode == CEED_TRANSPOSE);
+  const CeedScalar *u;
+  CeedScalar *v;
+  if (U != CEED_VECTOR_NONE) {
+    ierr = CeedVectorGetArrayRead(U, CEED_MEM_HOST, &u); CeedChkBackend(ierr);
+  } else if (eval_mode != CEED_EVAL_WEIGHT) {
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_BACKEND,
+                     "An input vector is required for this CeedEvalMode");
+    // LCOV_EXCL_STOP
+  }
+  ierr = CeedVectorGetArrayWrite(V, CEED_MEM_HOST, &v); CeedChkBackend(ierr);
+
+  // Clear v if operating in transpose
+  if (t_mode == CEED_TRANSPOSE) {
+    const CeedInt v_size = num_elem*num_comp*num_nodes;
+    for (CeedInt i = 0; i < v_size; i++)
+      v[i] = (CeedScalar) 0.0;
+  }
+  bool tensor_basis;
+  ierr = CeedBasisIsTensor(basis, &tensor_basis); CeedChkBackend(ierr);
+  // Tensor basis
+  if (tensor_basis) {
+    return CeedError(ceed, CEED_ERROR_BACKEND,
+                     "H(div) is not supported for tensor basis");
+  } else {
+    // Non-tensor basis H(div)
+    switch (eval_mode) {
+    // Interpolate to/from quadrature points
+    case CEED_EVAL_INTERP: {
+      CeedInt P = num_nodes, Q = dim*num_qpts;
+      const CeedScalar *interp;
+      ierr = CeedBasisGetInterp(basis, &interp); CeedChkBackend(ierr);
+      if (t_mode == CEED_TRANSPOSE) {
+        P = dim*num_qpts; Q = num_nodes;
+      }
+      ierr = CeedTensorContractApply(contract, num_comp, P, num_elem, Q,
```

Comment:
Does this support multiple components? Should it?

Reply:
Because of the way we created the restriction from plex, we set num_comp=1. It is like we have 1 vector dof, and the components of this dofs goes to interpolation matrix (dim*QxP, as I mentioned above)

Reply:
But is it possible mathematically to have a multiple components on a single Hdiv element? That feels like a natural possibility

Reply:
They really have multiple components, dof is a vector. At the end we use the H(div) basis to interpolate a vector variable (like velocity). But can we handle multiple components using tensor product as we do for H^1?! I don't know how.! 
For this element I implemented, we have P=8 values for x and P=8 values for y at quadrature points.

Reply:
I understand that at the quadrature points there are dim components. That's not what I mean. I mean that it feels natural that we should be able to represent multiple Hdiv fields of the same order with the same basis. Doing one interpolation with three fields seems it would be more effecient than three separate interpolations, for instance.

Reply:
Hmm, I see what you mean. So like what we have in H^1, we have one interpolation but use it for all other components (to handle vector values), right?

Reply:
Exactly.

 Right now the ref code uses ncamp in some places but not others, which is why I noticed it and asked.

Reply:
Yeah, there are methods for elasticity in which a tensor (stress or strain, without symmetry) lives in an H(div) space. From https://doi.org/10.1007/s10915-010-9373-2
![image](https://user-images.githubusercontent.com/3303/150704511-5d368e97-3fdb-4fcc-8c96-9aa9229be2ff.png)

Reply:
To my knowledge at least for this element (BDM element) I implemented, I don't know how to do that.
Perhaps we need to check `FEnics` to see how they handle it. https://launchpadlibrarian.net/83776282/fenics-book-2011-10-27-final.pdf.

Reply:
To make an H(div) space for matrices by treating each row as a standard H(div) space? I think your code would almost work, with `ncomp=dim`.

Reply:
There are layers of abstraction here. LibCEED is only responsible for the algebraic operation. As far as I'm concerned, it doesn't need to know anything about H(div) or H^1, it just has a matrix of shape `(dim*Q, P)` that may be Kroneckered with an identity to transform a vector of length `c*P` to a vector of length `c*dim*Q`. We want the basis to *also* be able to transform the vector of length `c*P` to `c*Q`. These operations have names, but perhaps there's nothing special about the operation (except internal structure most relevant to tensor elements).

Let k be the number of Q-vec components in each E-vec dof, and c be the number of E-vec components.

| operation | basis data shape | Q-vec shape | E-vec shape | notes |
|---|---|---|---|---|
| interp | (k, Q, P) | (c, k, Q) | (c, P) |
| grad | (dim, k, Q, P) | (dim, c, k, Q) | (c, P) |
| div | (Q, P) | (c, Q) | (c, P) | requires k=dim |
| curl(dim=2) | (1, Q, P) | (1, c, Q) | (c, P) | requires k=dim |
| curl(dim=3) | (dim, Q, P) | (dim, c, Q) | (c, P) | requires k=dim |

We have Poisson (k=1, c=1), standard elasticity (k=1, c=3), mixed/Darcy (k=dim, c=1), mixed elasticity/weak symmetry (k=dim, c=3), Nedelec for Maxwell (k=3, c=1). The "requires" notes are for that operation to be compatible with the corresponding interpolation. But perhaps someone will conceive of another operation so maybe we can make it possible to define new operations via their shape semantics. (Not in this PR.)

Reply:
I could be wrong, but it seems like from this chart we could have the basis separately track c and k to reduce repetition in the ref apply code and provide a simpler path forward to extending this to the GPU backends.

Reply:
> 

Yeah, that looks great. So we need to update `CeedTensorContractApply` function to take `k` as an input as well, right?

Reply:
The tensor contract shouldn't care. I think it would be a property of the basis during BasisApply.

Reply:
Ah, right. So we add this to the basis and for example set Q = k*num_qpts, in BasisApply. Like what we have here (dim*num_qpts).

Reply:
Right, that's roughly what I'm thinking. Though I would keep Q still as the number of qpts and just scale pre/post as needed by k.

Reply:
How about here, I think it is easy to add `k` to the basis constructor and set it to 1 for H^1 and `dim` for H(div), so I don't need to create `CeedBasisApplyHdiv_Ref`.

Reply:
Agreed. I think we had the same idea/comment at the same time in two different threads

Reply:
What would this k be called? `CeedBasisGetNumQuadratureComponents`?

Reply:
Isn't `Q-vec` component?

Reply:
Perhaps add `CeedBasisGetQVectorComponents`?

Reply:
QVector is just short for QuadraturePointVector

Reply:
Thanks, so what we should call it? `NumQuadratureComponenets` is a little confusing to me.

Reply:
Names are hard. I don't have a good  idea here. My thought of 

`CeedBasisGetNumQuadratureComponents` was from the discussion above
>Let k be the number of Q-vec components in each E-vec dof, and c be the number of E-vec components.

But I'm open to other ideas

Reply:
I added `Q_comp` and `CeedBasisGetNumQuadratureComponents` so we don't need `BasisApplyHdiv_Ref`.